wandb/sdk/system/assets/ipu.py

```diff
@@ -0,0 +1,167 @@
+                key = f"{key} ({suffix})"
+
+        try:
+            float_value = float(value)
+            num_value = int(float_value) if float_value.is_integer() else float_value
+        except ValueError:
+            return None
+
+        return key, num_value
+
+    def sample(self) -> None:
+        try:
+            stats = {}
+            devices = self._gc_ipu_info.getDevices()
+
+            for device in devices:
+                device_metrics: Dict[str, str] = dict(device)
+
+                pid = device_metrics.get("user process id")
+                if pid is None or int(pid) != self._pid:
+                    continue
+
+                device_id = device_metrics.get("id")
+                initial_call = device_id not in self._devices_called
+                if device_id is not None:
+                    self._devices_called.add(device_id)
+
+                for key, value in device_metrics.items():
+                    log_metric = initial_call or key in self.variable_metric_keys
+                    if not log_metric:
+                        continue
+                    parsed = self.parse_metric(key, value)
+                    if parsed is None:
+                        continue
+                    parsed_key, parsed_value = parsed
+                    stats[self.name.format(device_id, parsed_key)] = parsed_value
+
+            self.samples.append(stats)
+
+        except Exception as e:
+            wandb.termwarn(f"IPU stats error {e}", repeat=False)
+
+    def clear(self) -> None:
+        self.samples.clear()
+
+    def serialize(self) -> dict:
+        if not self.samples:
+            return {}
+        stats = {}
+        for key in self.samples[0].keys():
+            samples = [s[key] for s in self.samples if key in s]
+            aggregate = round(sum(samples) / len(samples), 2)
+            stats[key] = aggregate
+        return stats
+
+
+@asset_registry.register
+class IPU:
+    def __init__(
+        self,
+        interface: "Interface",
+        settings: "SettingsStatic",
+        shutdown_event: mp.synchronize.Event,
+    ) -> None:
+        self.name = self.__class__.__name__.lower()
+        self.metrics: List[Metric] = [
+            IPUStats(settings._stats_pid),
+        ]
+        self.metrics_monitor = MetricsMonitor(
+            self.metrics,
+            interface,
+            settings,
+            shutdown_event,
+        )
+
+    @classmethod
+    def is_available(cls) -> bool:
+        try:
+            import gcipuinfo  # type: ignore  # noqa: F401
+        except ImportError:
+            return False
+
+        return True
+
+    def start(self) -> None:
+        self.metrics_monitor.start()
+
+    def finish(self) -> None:
+        self.metrics_monitor.finish()
+
+    def probe(self) -> dict:
+        num_devices = len(self.metrics[0]._gc_ipu_info.getDevices())  # type: ignore
+        return {"ipu": {"device_count": num_devices, "vendor": "Graphcore"}}
```

Comment:
we probably can make it more meaningful given the comment above that some metrics are static?

Reply:
👍🏻 done